### PR TITLE
Initial implementation of the GitHub Action

### PR DESCRIPTION
**Pull request by sryabkov**
# what

Initial implementation of the GitHub Action to load .env files and export the key-value pair as environment variables with optional masking

# why

This PR was triggered by the fact that I liked [koheing/set-masked-env] but it wasn't parsing the .env files correctly in all cases (e.g. it was not stripping away comments). The original action was also using the old version of Node. 

I also thought I could use some JavaScript/TypeScript practice. 

# notes

This action can be used instead of [koheing/set-masked-env]. It adds an optional `removeQuotes` parameter. Notice that the default value for `mask` is `false` in this PR and `true` in [koheing/set-masked-env]

This action uses the [parse-dotenv](https://github.com/luqmanoop/parse-dotenv) package for parsing .env files

# todo

- [x] update README.md
- [ ] update tests

[koheing/set-masked-env]: https://github.com/koheing/set-masked-env

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -1,24 +1,25 @@
...
-    const ms: string = core.getInput('milliseconds')
+    const filePath: string = core.getInput('filePath')
+    const mask: boolean = core.getBooleanInput('mask')
+    const removeQuotes: boolean = core.getBooleanInput('removeQuotes')
 
     // Debug logs are only output if the `ACTIONS_STEP_DEBUG` secret is true
-    core.debug(`Waiting ${ms} milliseconds ...`)
+    core.debug(`filePath: ${filePath}`)
+    core.debug(`mask: ${mask}`)
+    core.debug(`mask: ${removeQuotes}`)
```
Comment (Jrfidellis) [inline src/main.ts:19]:
```suggestion
    core.debug(`removeQuotes: ${removeQuotes}`)
```

File: src/exportVars.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        }
+        core.exportVariable(key, finalValue)
+      }
+      resolve()
+    } catch (error) {
+      if (error instanceof Error)
+        throw new Error(`Error setting environment variables: ${error.message}`)
+    }
+  })
+}
```
Comment (Jrfidellis) [inline src/exportVars.ts:38]: This function doesn't need to be async because it's not waiting on any promises (all of the code here is synchronous). This can be noticed because there are no `promise.then` or `await promise`

Suggestion: 
```suggestion
export function exportVariables(
  entries: EnvObject,
  mask: boolean,
  removeQuotes: boolean
) {
    try {
      for (const [key, value] of Object.entries(entries)) {
        let finalValue = value
        if (
          removeQuotes &&
          ((value.startsWith('"') && value.endsWith('"')) ||
            (value.startsWith("'") && value.endsWith("'")))
        ) {
          finalValue = value.slice(1, -1)
        }
        if (mask) {
          core.setSecret(finalValue)
        }
        core.exportVariable(key, finalValue)
      }
    } catch (error) {
      if (error instanceof Error)
        throw new Error(`Error setting environment variables: ${error.message}`)
    }
}
```

File: src/readFile.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  return new Promise(resolve => {
+    try {
+      const fileEntries: EnvObject = parseEnv(filePath)
+      resolve(fileEntries)
+    } catch (error) {
+      if (error instanceof Error)
+        throw new Error(`Error parsing file ${filePath}: ${error.message}`)
+    }
+  })
+}
```
Comment (Jrfidellis) [inline src/readFile.ts:19]: This function doesn't need to be async because it's not waiting on any promises (all of the code here is synchronous).

Suggestion:
```suggestion
export function getFileEntries(filePath: string): EnvObject {
    try {
      const fileEntries: EnvObject = parseEnv(filePath)
      return fileEntries
    } catch (error) {
      if (error instanceof Error)
        throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
}
```

Comment (sryabkov) [reply src/readFile.ts:19]: @Jrfidellis Is there harm in this function being a promise? Here is why I am asking... 

If we go with the the suggested non-async alternative, we actually need to make some changes to it. The suggested code right now fails checks with 

```
Function lacks ending return statement and return type does not include 'undefined'.
```

So, we either need to change the return type to `EnvObject | undefined` and then propagate this type change to the caller and possibly beyond paying attention to where we need to check for undefined 

```
export function getFileEntries(filePath: string): EnvObject | undefined {
  try {
    const fileEntries: EnvObject = parseEnv(filePath)
    return fileEntries
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
  }
}
```

Or create an empty object upfront and potentially return an empty object to the caller and deal with that in the caller. 

```
export function getFileEntries(filePath: string): EnvObject {
  let fileEntries: EnvObject = {}
  try {
    fileEntries = parseEnv(filePath)
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Error parsing file ${filePath}: ${error.message}`)
    }
  }
  return fileEntries
}
```

So, by simplifying on one place, we might be making things more complex in other places. Would you recommendation be different, if the calling code in `main.ts` be re-written with a promise chain?

Below is ChatGPTs answer to the question `should promises only be used for async code?`


```
Promises are a fundamental concept in JavaScript primarily used for handling asynchronous operations, but they are not limited to async code exclusively. Promises are designed to represent the eventual completion (or failure) of an asynchronous operation, providing a way to handle its result or error once it's available.

However, promises can also be used in synchronous code to manage flow control and improve readability, especially when dealing with sequences of operations where some may be asynchronous and others synchronous.

Here are some key points about promises:

**Asynchronous Operations**: Promises are most commonly used to handle asynchronous operations such as fetching data from a server, reading files, or any operation that takes time to complete and may not immediately return a result.

**Synchronous Operations**: Promises can also be useful in synchronous code. For example, if you have a sequence of synchronous operations followed by an asynchronous one, you can use promises to handle the result of the asynchronous operation in a structured way.

**Error Handling**: Promises provide a standardized way to handle both success and failure scenarios, regardless of whether the operation is asynchronous or synchronous. This makes error handling more consistent and manageable.

**Chaining**: Promises support method chaining (`then`, `catch`, `finally`), which can improve code readability and maintainability by allowing you to write sequential operations without nesting callbacks.

In conclusion, while promises are commonly associated with asynchronous code, they are a versatile tool that can also be applied to synchronous code to improve structure, readability, and error handling. They provide a standardized way to handle the outcome of operations, whether they are immediate or take time to complete.

```

I don't know what the right answer is, or even if there is the right answer, so I am just trying to learn.

File: src/exportVars.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        if (mask) {
+          core.setSecret(finalValue)
+        }
+        core.exportVariable(key, finalValue)
+      }
+      resolve()
+    } catch (error) {
+      if (error instanceof Error)
+        throw new Error(`Error setting environment variables: ${error.message}`)
+    }
```
Comment (Ikstar) [inline src/exportVars.ts:36]: Should always put an else clause, and throw the original if we don't know what got thrown. Otherwise this thin will silently eat errors.

**Review by Ikstar — APPROVED** [review]
Looks good overall, I agree with Joao's comments, this is all synchronous by default.

Next iteration we may want to convert empty strings
e.g. `ENV_VAR_1=` so that ENV_VAR_1 is `null`

File: src/readFile.ts (hunk 3 above)
Comment (JoaoLSS) [reply src/readFile.ts:19]: it asks for `undefined` as the return type because of the conditional throw inside the catch block, if for some reason it falls into the catch block but does not re-throw, it implicit returns `undefined`, one quick fix is to throw an `unknown error` if the error is not an instance of `Error`

Comment (sryabkov) [reply src/readFile.ts:19]: Yes, but then the function still returns either an `EnvObject` or an `Error`, which is two different datatypes, and the caller, strictly speaking, needs to check. I was just wondering if making every function a promise and attaching `.catch` at the end of the promise chain makes error-handling easier...

File: src/exportVars.ts (hunk 2 above)
Comment (sryabkov) [reply src/exportVars.ts]: Changed in f77cf96d07e31eaecf2cb45c6fcc4933c2c4c538

File: src/readFile.ts (hunk 3 above)
Comment (sryabkov) [reply src/readFile.ts]: With changed error handling, the function now either returns an `EnvObject` or throws an exception. See f77cf96d07e31eaecf2cb45c6fcc4933c2c4c538

File: src/exportVars.ts (hunk 4 above)
Comment (sryabkov) [reply src/exportVars.ts]: I believe this has been addressed in f77cf96d07e31eaecf2cb45c6fcc4933c2c4c538
